README.rst
######
mctele
######

|License: MIT| |Build status|

A simple Telegram bot that uses the `Minecraft Query protocol
<https://wiki.vg/Query>`_ to get the list of players on a server, and
sends changes in the list to a Telegram chat.

Usage
=====

To use, simply run the built executable with the following environment
variables set:

* ``MCTELE_BOT_TOKEN``: Telegram bot token
* ``MCTELE_CHAT_ID``: Chat ID to send updates to
* ``MCTELE_SERVER_ADDR``: (optional) Address with port to query (e.g.
  ``192.168.1.1:25565``. Defaults to ``127.0.0.1:25565``.
* ``MCTELE_QUERY_INTERVAL``: (optional) Interval in seconds between each
  query to make. Defaults to 15 seconds.
* ``MCTELE_KEEP_OLD``: (optional) If set, keep old messages when new one
  is sent. Otherwise, delete the old message.
* ``MCTELE_SILENT``: (optional) If set, send messages with silent
  notifications.

Do not forget to set in the ``server.properties``::

	enable-query=true
	query.port=<something>

.. |License: MIT| image:: https://img.shields.io/badge/License-MIT-yellow.svg
	:target: https://opensource.org/licenses/MIT

.. |Build status| image:: https://github.com/chuahou/mctele/workflows/nix%20test/badge.svg?branch=master


CHANGELOG.md
# Revision history for mctele

## 0.3.0.0 -- 2021-01-08

* Modify formatting of messages.

## 0.2.0.0 -- 2021-01-08

* Delete old messages when new one is sent (unless `MCTELE_KEEP_OLD` is set).
* Send silent messages if `MCTELE_SILENT` is set.

## 0.1.0.0 -- 2021-01-06

* First version. Released on an unsuspecting world.


src/Mctele/Server.hs
-- SPDX-License-Identifier: MIT
-- Copyright (c) 2021 Chua Hou

module Mctele.Server ( getInfo
                     , ServerInfo
                     ) where

import qualified Control.Exception          as Exception
import           Data.Attoparsec.ByteString (Parser, anyWord8, manyTill,
                                             maybeResult, parse, string,
                                             takeWhile1, word8)
import           Data.ByteString            (ByteString, concat, pack)
import qualified Data.ByteString.Char8      as C8
import           Data.Functor               (($>))
import           Data.Int                   (Int32)
import           Data.Serialize             (encode)
import           Data.String.Interpolate    (i)
import           Data.Word                  (Word8)
import qualified Network.Socket             as Net
import qualified Network.Socket.ByteString  as Net
import           System.Timeout             (timeout)
import           Text.Read                  (readMaybe)

-- | Contains all information we need about a server. @Nothing@ if server is
-- offline; @Just@ list of player usernames otherwise.
type ServerInfo = Maybe [Player]

-- | Players are represented by their username string.
type Player = String

-- | Creates a packet with given type and payload.
makePacket :: Word8 -> ByteString -> ByteString
makePacket typ payload = Data.ByteString.concat
                            [ pack [ 0xFE, 0xFD ] -- magic prefix
                            , pack [ typ ]
                            , pack [ 0x00, 0x00, 0x00, 0x00 ] -- session ID
                            , payload
                            ]

-- | Given a socket address, get the server at that address's information.
getInfo :: Net.SockAddr -> IO ServerInfo
getInfo addr = tryGetInfo attempts
    where
        attempts  = 10 :: Int
        timeoutMs = 1000000

        tryGetInfo 0 = pure Nothing -- give up
        tryGetInfo n = timeout timeoutMs (Exception.try handshake
                        :: IO (Either Exception.IOException ServerInfo))
                     >>= maybe tryAgain (\case             -- timeout
                            Left _        -> tryAgain      -- IOException
                            Right Nothing -> tryAgain      -- unsuccessful query
                            Right info    -> succeed info) -- successful query
            where
                tryAgain =  putStrLn [i|Trying again, #{n - 1} attempt(s) left|]
                         >> tryGetInfo (n - 1)
                succeed = (putStrLn "Successful query" $>)

        handshake = do
            { sock <- Net.socket Net.AF_INET Net.Datagram Net.defaultProtocol
            ; Net.connect sock addr
            ; Net.sendAll sock (makePacket 0x09 (pack []))
            ; response <- Net.recv sock 128
            ; case (readMaybe (init . drop 5 . C8.unpack $ response)
                    :: Maybe Int32) of
                Just challenge -> request challenge sock <* Net.close sock
                Nothing        -> Nothing                <$ Net.close sock
            }

        request challenge sock = do
            { Net.connect sock addr
            ; Net.sendAll sock (makePacket 0x00 (Data.ByteString.concat
                                    [ encode challenge
                                    , pack (replicate 4 0x00) -- padding
                                    ]))
            ; response <- Net.recv sock 1024
            ; case maybeResult $ parse playersP response of
                Nothing -> pure Nothing
                Just ps -> pure $ Just ps
            }

        -- Parses full stats and returns list of players.
        playersP :: Parser [Player]
        playersP =  manyTill anyWord8 (string padding)
                 >> map C8.unpack <$>
                    manyTill (takeWhile1 (/= 0x00) <* word8 0x00) (word8 0x00)
            where
                padding = pack [ 0x01, 0x70, 0x6C, 0x61, 0x79
                               , 0x65, 0x72, 0x5F, 0x00, 0x00
                               ]


src/Mctele/Telegram.hs
-- SPDX-License-Identifier: MIT
-- Copyright (c) 2021 Chua Hou

{-# LANGUAGE DeriveGeneric #-}

module Mctele.Telegram ( BotToken
                       , ChatID
                       , MessageID
                       , sendServerStatus
                       , deleteMessage
                       ) where

import           Mctele.Server

import           Control.Monad           (void, (>=>))
import           Data.Aeson              (FromJSON)
import           Data.ByteString.Char8   (pack)
import           Data.Functor            ((<&>))
import           Data.List               (intercalate)
import           Data.String.Interpolate (i)
import           GHC.Generics            (Generic)
import           Network.HTTP.Simple     (Request, getResponseBody,
                                          httpJSONEither, httpNoBody,
                                          parseRequest, setRequestQueryString)

type BotToken  = String
type ChatID    = String
type MessageID = String

-- | @createRequest tok op params@ creates a Telegram API request with bot token
-- @tok@, operation @op@ and tuples of @(key, value)@s as parameters @params@.
createRequest :: BotToken -> String -> [(String, String)] -> IO Request
createRequest tok op params = setRequestQueryString query <$> baseRequest
    where
        baseRequest    = parseRequest
                            [i|https://api.telegram.org/bot#{tok}/#{op}|]
        query          = map mkParam params
        mkParam (k, v) = (pack k, pure . pack $ v)

-- | @sendServerStatus silent tok chatId info@ uses the Telegram Bot API with
-- token @tok@ to send server status @info@ to chat with ID @chatId@, with
-- silent notifications if @silent@ is true.
--
-- Returns the resulting message ID.
sendServerStatus :: Bool -> BotToken -> ChatID -> ServerInfo -> IO (Maybe MessageID)
sendServerStatus silent tok chatId info =   request >>= httpJSONEither
                                        <&> either (const Nothing) getMessageId
                                        .   getResponseBody
    where
        request = createRequest tok "sendMessage" $
                    [ ("chat_id",    chatId)
                    , ("text",       text info)
                    , ("parse_mode", "MarkdownV2")
                    ] ++ [("disable_notification", "true") | silent]
        text (Just []) =   "\\[Server\\] *ONLINE*"
        text (Just ps) = [i|\\[Server\\] *ONLINE*, #{length ps} players: `#{intercalate "`, `" ps}`|]
        text Nothing   =   "\\[Server\\] *OFFLINE*"

-- | Deletes the specified message on Telegram.
deleteMessage :: BotToken -> ChatID -> MessageID -> IO ()
deleteMessage tok chatId msgId = request >>= void . httpNoBody
    where
        request = createRequest tok "deleteMessage"
                    [ ("chat_id",    chatId)
                    , ("message_id", msgId)
                    ]

-- | Returns 'Just' the message ID for a successful request and 'Nothing'
-- otherwise, with the request response JSON as input.
getMessageId :: TelegramResponse -> Maybe MessageID
getMessageId = result >=> (fmap show . message_id)

-- | Telegram API request response JSON type. We are only interested in the
-- result member.
newtype TelegramResponse = TelegramResponse { result :: Maybe TelegramResult }
    deriving Generic
instance FromJSON TelegramResponse

-- | Telegram API request response result JSON type. We are only interested in
-- the resulting message ID.
newtype TelegramResult = TelegramResult { message_id :: Maybe Int }
    deriving Generic
instance FromJSON TelegramResult


exe/Main.hs
-- SPDX-License-Identifier: MIT
-- Copyright (c) 2021 Chua Hou

module Main where

import           Mctele.Server
import           Mctele.Telegram

import           Control.Concurrent           (threadDelay)
import           Control.Monad                (forM_)
import           Data.Char                    (isDigit)
import           Data.Functor                 ((<&>))
import           Data.Maybe                   (fromMaybe, isJust)
import qualified Network.Socket               as Net
import           System.Environment           (getEnv, lookupEnv)
import           Text.ParserCombinators.ReadP
import           Text.Read                    (readMaybe)

main :: IO ()
main = do
    -- Telegram bot token
    { tok      <- getEnv    "MCTELE_BOT_TOKEN"

    -- Telegram chat ID
    ; chatId   <- getEnv    "MCTELE_CHAT_ID"

    -- Address of minecraft server with port
    ; addr     <- lookupEnv "MCTELE_SERVER_ADDR"    <&> maybe localhost mkAddr

    -- Interval in seconds between queries
    ; interval <- lookupEnv "MCTELE_QUERY_INTERVAL" <&> fromMaybe 15
                                                    . (>>= readMaybe)

    -- Whether to keep old messages (delete otherwise)
    ; keepOld  <- lookupEnv "MCTELE_KEEP_OLD"       <&> isJust

    -- Whether to send messages with silent notifications
    ; silent   <- lookupEnv "MCTELE_SILENT"         <&> isJust

    ; loop tok chatId addr interval keepOld silent Nothing Nothing
    }
    where
        localhost :: Net.SockAddr
        localhost = Net.SockAddrInet 25565 (Net.tupleToHostAddress
                                                (127, 0, 0, 1))

        mkAddr :: String -> Net.SockAddr
        mkAddr cs = case readP_to_S addrP cs of
                      [(y, "")] -> y
                      _         -> error "Invalid server address"

        addrP :: ReadP Net.SockAddr
        addrP = do
            { n1   <- numberP <* char '.'
            ; n2   <- numberP <* char '.'
            ; n3   <- numberP <* char '.'
            ; n4   <- numberP <* char ':'
            ; port <- numberP <* eof
            ; pure $ Net.SockAddrInet port (Net.tupleToHostAddress
                                                (n1, n2, n3, n4))
            }
            where
                numberP :: Read a => ReadP a
                numberP =   many1 (satisfy isDigit)
                        >>= \cs -> case readMaybe cs of
                                     Just w  -> pure w
                                     Nothing -> fail "Invalid number"

        loop tok chatId addr interval keepOld silent = go
            where
                go prev prevMsgId = do
                    { info  <- getInfo addr
                    ; msgId <- if prev /= info
                                  then sendServerStatus silent tok chatId info
                                    <* forM_ prevMsgId (deleteMessage tok chatId)
                                  else pure prevMsgId
                    ; threadDelay (1000000 * interval)
                    ; go info (if keepOld then Nothing else msgId)
                        -- don't propagate message ID to delete if we want to
                        -- keep old messages
                    }
